algebra/Atiyah.lean

```diff
@@ -0,0 +1,1695 @@
+| h := ⟨λ x, calc
+  x = x * 1   : by simp
+  ... = x * 0 : by rw h
+  ... = 0     : by simp⟩
+
+
+-- page 2
+
+class is_hom {α : Type u} {β : Type v} [comm_ring α] [comm_ring β] (f : α → β) : Prop :=
+(map_add : ∀ {x y}, f (x + y) = f x + f y)
+(map_mul : ∀ {x y}, f (x * y) = f x * f y)
+(map_one : f 1 = 1)
+
+namespace is_hom
+
+variables {α : Type u} {β : Type v} [comm_ring α] [comm_ring β]
+variables (f : α → β) [is_hom f] {x y : α}
+
+attribute [simp] map_add
+attribute [simp] map_mul
+attribute [simp] map_one
+
+@[simp] lemma map_zero : f 0 = 0 :=
+calc f 0 = f (0 + 0) - f 0 : by rw [map_add f]; simp
+     ... = 0 : by simp
+
+@[simp] lemma map_neg : f (-x) = -f x :=
+calc f (-x) = f (-x + x) - f x : by rw [map_add f]; simp
+        ... = -f x : by simp [map_zero f]
+
+@[simp] lemma map_sub : f (x - y) = f x - f y :=
+by simp [map_add f, map_neg f]
+
+end is_hom
+
+class subring (α : Type u) [comm_ring α] (S : set α) : Prop :=
+(add_mem : ∀ {x y}, x ∈ S → y ∈ S → x + y ∈ S)
+(neg_mem : ∀ {x}, x ∈ S → -x ∈ S)
+(mul_mem : ∀ {x y}, x ∈ S → y ∈ S → x * y ∈ S)
+(one_mem : (1:α) ∈ S)
+
+namespace subring
+
+variables (α : Type u) [comm_ring α] (S : set α) [subring α S]
+
+instance : comm_ring S :=
+{ add            := λ ⟨x, hx⟩ ⟨y, hy⟩, ⟨x + y, add_mem hx hy⟩,
+  add_assoc      := λ ⟨x, hx⟩ ⟨y, hy⟩ ⟨z, hz⟩, subtype.eq $ add_assoc x y z,
+  zero           := ⟨0, eq.subst (add_neg_self (1:α)) $ add_mem (one_mem S) $ neg_mem $ one_mem S⟩,
+  zero_add       := λ ⟨x, hx⟩, subtype.eq $ zero_add x,
+  add_zero       := λ ⟨x, hx⟩, subtype.eq $ add_zero x,
+  neg            := λ ⟨x, hx⟩, ⟨-x, neg_mem hx⟩,
+  add_left_neg   := λ ⟨x, hx⟩, subtype.eq $ add_left_neg x,
+  add_comm       := λ ⟨x, hx⟩ ⟨y, hy⟩, subtype.eq $ add_comm x y,
+  mul            := λ ⟨x, hx⟩ ⟨y, hy⟩, ⟨x * y, mul_mem hx hy⟩,
+  mul_assoc      := λ ⟨x, hx⟩ ⟨y, hy⟩ ⟨z, hz⟩, subtype.eq $ mul_assoc x y z,
+  one            := ⟨1, one_mem S⟩,
+  one_mul        := λ ⟨x, hx⟩, subtype.eq $ one_mul x,
+  mul_one        := λ ⟨x, hx⟩, subtype.eq $ mul_one x,
+  left_distrib   := λ ⟨x, hx⟩ ⟨y, hy⟩ ⟨z, hz⟩, subtype.eq $ left_distrib x y z,
+  right_distrib  := λ ⟨x, hx⟩ ⟨y, hy⟩ ⟨z, hz⟩, subtype.eq $ right_distrib x y z,
+  mul_comm       := λ ⟨x, hx⟩ ⟨y, hy⟩, subtype.eq $ mul_comm x y }
+
+@[simp] lemma add (x y : α) (hx : x ∈ S) (hy : y ∈ S) :
+(⟨x, hx⟩ : S) + ⟨y, hy⟩ = ⟨x + y, add_mem hx hy⟩ := rfl
+
+@[simp] lemma mul (x y : α) (hx : x ∈ S) (hy : y ∈ S) :
+(⟨x, hx⟩ : S) * ⟨y, hy⟩ = ⟨x * y, mul_mem hx hy⟩ := rfl
+
+instance is_hom : is_hom ((λ x, x) : S → α) :=
+{ map_add := λ x y, by cases x; cases y; refl,
+  map_mul := λ x y, by cases x; cases y; refl,
+  map_one := rfl }
+
+end subring
+
+instance is_hom.comp (α : Type u) (β : Type v) (γ : Type w)
+[comm_ring α] [comm_ring β] [comm_ring γ]
+(g : β → γ) (f : α → β) [is_hom g] [is_hom f] : is_hom (g ∘ f) :=
+{ map_add := λ x y, by simp [function.comp, is_hom.map_add f, is_hom.map_add g],
```

Comment:
the second and third line should be indented.  Is it necessary that `α`, `β`, and `γ` are explict? I would prefer when you add them as {implict} variables at the start of your lean file.